### PR TITLE
fix: time.Time rendering in the diff returned by error message

### LINKED ISSUE #1078: Diffs on time.Time objects nested inside structs produce hard to read diffs
We recently observed tough to read diffs produced by testify on assertion failures on structs when the structs contain nested `time.Time` objects.

I have created this repo with this testfile that does a minimal repro of the issue https://github.com/shawc71/testify-issue-demo/blob/master/example_test.go

In summary, given a struct like:

```
type structWithTime struct {
        	someTime   *time.Time
        	someString string
}
```

the following produces a [diff that is 1220 lines+](https://github.com/stretchr/testify/files/6527166/test_output.txt) and unparsable for humans:

```
		expectedTime, err := time.Parse("2006-01-02", "2020-05-22")
		assert.NoError(t, err)
		expected := &structWithTime{
			someTime:   &expectedTime,
			someString: "some val",
		}

		loc, err := time.LoadLocation("America/New_York")
		assert.NoError(t, err)
		actualTime, err := time.ParseInLocation("2006-01-02", "2020-05-21", loc)
		actual := &structWithTime{
			someTime:   &actualTime,
			someString: "some val",
		}
		assert.Equal(t, expected, actual)
```
      
I believe this was an unintended side-effect of https://github.com/stretchr/testify/pull/895. 

This makes doing assertions on structs with time.Time pretty rough. I imagine having a time.Time in a struct and being able to do assertions on those structs is a pretty common use case so I think we should make this nicer. :)

I am working on a PR to address this :)

### PR DESCRIPTION
**Pull request by ccoVeille**
## Summary

Improve formatting when dealing with time.Time now we vendored gospew

## Changes

Now, go-spew is vendored, we can fix old issues we had.

We can fix the representation of [time.Time](https://pkg.go.dev/time#Time) in gospew

And remove the imperfect fix we had in place.

## Motivation

#895 was created by @luan [to solve some legitimate issues](https://github.com/stretchr/testify/pull/895#issue-568723418). It was released with [v1.6.0](https://github.com/stretchr/testify/releases/tag/v1.6.0) on May 29, 2020

- #895 

But unfortunately, the changes lead to an issue with the way [time.Time](https://pkg.go.dev/time#Time) are rendered. So #989  was opened on Aug 1, 2020 by @AlekSi

- #989

The issue was somehow fixed with #1072 by @HaraldNordgren 

- #1072 

#1072 was merged in Apr 27, 2021 (but only released in Mar 1, 2024 with[ v1.7.1](https://github.com/stretchr/testify/releases/tag/v1.7.1) )

Unfortunately, the fix was incomplete, as it was working only for first level, so #1078 was opened on May 22, 2021 by @shawc71

- #1078

#1079 was opened by @shawc71 on May 22, 2021 to try to fix the issue

- #1079

But it was never merged, as the solution was also imperfect, and because they were no solution inside testify to fix this

So #1078 was left open until now.

## Related issues

- Related to #1826
- Closes #1079
- Closes #1078

### CODE REVIEW

**Comment by ccoVeille** [comment]
@brackendawson please review, it's ready.

I'm facing this issue with testify for ages, I cannot wait to see it fixed :grimacing:

**Comment by fredbi** [comment]
@ccoVeille  I found your PR definitely interesting but perhaps lacking a few things.
You may find an alternate implementation there https://github.com/go-openapi/testify/pull/27
The main differences are:
* does not use type string representation, but type convertibility to time
* does not sort using the string representation (time.Time does **NOT** implement sort.Interface), but uses time.Time.Compare instead
* more tests with embedded types, private fields, pointers, nil values etc

Could be interesting to exchange our views

**Comment by ccoVeille** [comment]
Thanks @fredbi. I will take a look.

I saw the PRs you made in the last days.

I agree with you about the string sort that can lead to problem when comparing for example 999-12-31 and 1000-01-01

I'll review the PR in your fork also

File: internal/spew/common.go [hunk 1]
```diff
@@ -271,6 +271,12 @@ func canSortSimply(kind reflect.Kind) bool {
 	case reflect.Array:
 		return true
 	}
+
+	if kind == reflect.Struct && kind.String() == "time.Time" {
+		// time.Time values can be sorted directly, as it implements [sort.Interface].
+		return true
+	}
+
```
[the remark this answers is not in the document]
Comment (ccoVeille) [reply internal/spew/common.go]: Copilot got a point

File: internal/spew/format.go [hunk 2]
```diff
@@ -222,7 +222,7 @@ func (f *formatState) format(v reflect.Value) {
 
 	// Call Stringer/error interfaces if they exist and the handle methods
 	// flag is enabled.
-	if !f.cs.DisableMethods {
+	if !f.cs.DisableMethods || (f.cs.EnableTimeStringer && v.Type().String() == "time.Time") {
```
[the remark this answers is not in the document]
Comment (ccoVeille) [reply internal/spew/format.go]: Copilot got a point

File: internal/spew/spew_test.go [hunk 3]
```diff
@@ -203,6 +207,11 @@ func initSpewTests() {
 		{scsNoPtrAddr, fCSSdump, "", tptr, "(*spew_test.ptrTester)({\ns: (*struct {})({\n})\n})\n"},
 		{scsNoCap, fCSSdump, "", make([]string, 0, 10), "([]string) {\n}\n"},
 		{scsNoCap, fCSSdump, "", make([]string, 1, 10), "([]string) (len=1) {\n(string) \"\"\n}\n"},
+
+		// time.Time formatting:
+		{scsDefault, fCSFprint, "", tm, "2006-01-02 15:04:05.999999999 +0000 UTC"},
+		{scsNoMethods, fCSFprint, "", tm, "{999999999 63271811045 <nil>}"},
```
[the remark this answers is not in the document]
Comment (ccoVeille) [reply internal/spew/spew_test.go]: True

**Comment by ccoVeille** [comment]
Copilot provided the same feedback as you @fredbi 

This time I got it, haha

**Comment by fredbi** [comment]
> Copilot provided the same feedback as you @fredbi
> 
> This time I got it, haha

Should I feel flattered?  Mmh not so sure :)
Anyway you may always take a look at the implementation at github.com/go-openapi/testify (v2). This part has been merged long ago. All further improvements there are of course welcome.

**Comment by ccoVeille** [comment]
> > Copilot provided the same feedback as you @fredbi
> > 
> > This time I got it, haha
> 
> Should I feel flattered?  Mmh not so sure :)

Haha. It's simply that:
- I haven't check last time
- Copilot spotted the exact lines that are bad. It was easier to figure out what was the problem.


> Anyway you may always take a look at the implementation at github.com/go-openapi/testify (v2). This part has been merged long ago. All further improvements there are of course welcome. 

I'm already looking at it.
